modules/azure/nsg.go

```diff
@@ -0,0 +1,194 @@
+
+	// Get an authorizer
+	auth, err := NewAuthorizer()
+	if err != nil {
+		return network.DefaultSecurityRulesClient{}, err
+	}
+
+	nsgClient.Authorizer = *auth
+	return nsgClient, nil
+}
+
+// GetCustomNsgRulesClientE returns a rules client
+func GetCustomNsgRulesClientE(subscriptionID string) (network.SecurityRulesClient, error) {
+	// Validate Azure subscription ID
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return network.SecurityRulesClient{}, err
+	}
+
+	nsgClient := network.NewSecurityRulesClient(subscriptionID)
+
+	// Get an authorizer
+	auth, err := NewAuthorizer()
+	if err != nil {
+		return network.SecurityRulesClient{}, err
+	}
+
+	nsgClient.Authorizer = *auth
+	return nsgClient, nil
+}
+
+// GetAllNSGRulesE returns a slice containing all rules from the NSG (including defauls and custom)
+func GetAllNSGRulesE(resourceGroupName, nsgName, subscriptionID string) (NsgRuleSummaryList, error) {
+	defaultRulesClient, err := GetDefaultNsgRulesClientE(subscriptionID)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	customRulesClient, err := GetCustomNsgRulesClientE(subscriptionID)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	defaultRuleList, err := defaultRulesClient.ListComplete(context.Background(), resourceGroupName, nsgName)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	customRuleList, err := customRulesClient.ListComplete(context.Background(), resourceGroupName, nsgName)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	rules := make([]NsgRuleSummary, 0)
+	bindRuleList(&rules, defaultRuleList)
+	bindRuleList(&rules, customRuleList)
+	ruleList := NsgRuleSummaryList{}
+	ruleList.SummarizedRules = rules
+	return ruleList, nil
+}
+
+// NsgRuleSummaryList holds a colleciton of NsgRuleSummary structs
+type NsgRuleSummaryList struct {
+	SummarizedRules []NsgRuleSummary
+}
+
+// NsgRuleSummary is a string-based summary of an NSG rule with methods attached.
+type NsgRuleSummary struct {
+	Name                     string
+	Description              string
+	Protocol                 string
+	SourcePortRange          string
+	DestinationPortRange     string
+	SourceAddressPrefix      string
+	DestinationAddressPrefix string
+	Access                   string
+	Priority                 int32
+	Direction                string
+}
+
+func bindRuleList(dest *[]NsgRuleSummary, source network.SecurityRuleListResultIterator) {
+	for source.NotDone() {
+		v := source.Value()
+		*dest = append(*dest, convertToNsgRuleSummary(v.Name, v.SecurityRulePropertiesFormat))
+		err := source.NextWithContext(context.Background())
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func convertToNsgRuleSummary(name *string, rule *network.SecurityRulePropertiesFormat) NsgRuleSummary {
+	summary := NsgRuleSummary{}
+
+	if rule.Description != nil {
+		summary.Description = *rule.Description
+	} else {
+		summary.Description = ""
+	}
```

Comment:
NIT:  empty string is the default value anyway if not set, so you could remove the `else` clause.

Reply:
Updated default handling to safe pointer helpers.